### PR TITLE
Cache the list of emojis between company complete calls

### PR DESCRIPTION
**Pull request by torgeir**
Cache the list of emojis between company complete calls to speed up completion.

### CODE REVIEW

**Comment by dunn** [comment]
Oh, thanks!  Caching them is a great idea.  My one concern is that if a user modifies the `company-emoji-aliases` custom variable, those won't take effect until Emacs is restarted.  Any idea how we might get around that?

**Comment by torgeir** [comment]
Oh, I didn't realize that. Detecting a defcustom change isn't possible afaik.. We could expose a function to add aliases and regenerate the list when its done? And change the cached list to a `defvar`

**Comment by dunn** [comment]
Sorry for the delay in replying!  That sounds promising, if you want to try making that change.

**Comment by torgeir** [comment]
I won't have time to have a go at this in the near future, feel free to close it

File: company-emoji.el [hunk 1]
```diff
@@ -117,9 +117,11 @@
 (require 'company)
 (require 'company-emoji-list)
 
-;; version string
+(defconst company-emoji-version "2.5.0"
+  "Current version of company-emoji.")
 
-(defconst company-emoji-version "2.5.0")
+(defconst emojis (company-emoji-list-create)
```
Comment (wyuenho) [inline company-emoji.el]: You need to prefix this to prevent namespace collision

https://github.com/bbatsov/emacs-lisp-style-guide#naming

**Comment by wyuenho** [comment]
@dunn  You can use a `:set` keyword to recreate the list of emojis when the aliases have changed.
https://www.gnu.org/software/emacs/manual/html_node/elisp/Variable-Definitions.html

**Comment by torgeir** [comment]
So this took a while :grin:

The last commit prefixes the cached list of emojis and uses `:set` as per @wyuenho's suggestion to recreate the cached list on `company-emoji-aliases` change.

**Comment by torgeir** [comment]
And on second thoughts I really don't see how this is nescessary as the aliases are looked up at runtime when invoking `company-emoji`, before the list of emojis is consulted.

https://github.com/dunn/company-emoji/blob/master/company-emoji.el#L192-L193

**Comment by torgeir** [comment]
Yeah, it has no effect. Reverted it to keep only the namespaced var. Seems to work fine on my end, with added user aliases appearing instantly.

**Review by dunn — APPROVED** [review]

**Comment by dunn** [comment]
Thanks so much!

**Pull request merged**